Implementation of a new function to rotate an element

SebastienJoly opened the pull request:
Hello,

I tried to solve the issue with the dipole (or any element) rotation in pyAT. From what I have understood from the past issues on github and some tests I have done, the rotation transformation only changed the R1 and R2 attributes and left T1 and T2 untouched. This is true for a tilt/roll angle but not for a pitch or yaw angle.

I tried to reimplement the rotation function based on the Euler angles (https://en.wikipedia.org/wiki/Euler_angles) and this paper from ALS (https://www.sciencedirect.com/science/article/pii/S0168900222007793) on which is based the SC/pySC equivalent function. I put as many comments as possible to guide the user in understanding how the rotations are applied.

Now, the current function still returns non commutative rotations but agree perfectly with pySC. I used the MLS as an example where I rotated a quadrupole and observed the resulting closed-orbit.

Roll angle:
![roll](https://github.com/user-attachments/assets/4ab6a0ba-9cc1-4e09-ae25-00b7b03dcde5)

Yaw angle:
![yaw](https://github.com/user-attachments/assets/cc7c149a-aaf4-42d4-a5fb-719950f1bd31)

Pitch angle:
![pitch](https://github.com/user-attachments/assets/e1dec5e4-ce36-467f-9a06-2eb83d7c965e)

We can clearly see the differences w.r.t the old implementation. I also tried it with a dipole and it agreed with the pySC curves. In that case, I also account for the bending angle in the R1, R2, T1 and T2 calculations.

I think it would be relevant to move the rotation functions outside of the utils.py module and create a rotation.py one. What do you think?

@TeresiaOlsson and I will compare these results with AT and SC next.

lfarv commented:
@SebastienJoly : very nice!
- I fully agree with the idea of moving rotation functions out of `utils.py`,
- I do not know what to do with the old implementation: you are keeping both side by side, and the name of the new implementation (`rotate_translate_elem`) is puzzling for me. And the global `set_rotation` function will still use the old implementation. Otherwise we could simply replace the old function by the new one. Or another possibility is to move the old implementation to a `deprecated` module or package: if necessary, it could be used with `from at.deprecated import rotate_elem`,
- I think that there is a problem related to AT's choice of coordinates (also present in the old implementation): T1[1] and T1[3] are used to model rotations around the X or Y axis. But these coordinates are $p_x= x' (1+\delta)$ and not $x'$: so the angular kick $x'$ depends on the momentum. No way to deal with that with a matrix… This should introduce an error on the dispersion generated by rotation errors. Since SC/pySC is based AT for the optics, I don't think comparing the dispersion will show anything. Is there another possible  benchmark?

oscarxblanco commented:
Dear @SebastienJoly , thank you very much. 
The documentation still misses the reference point around which the transformation is applied for curved reference systems.
Or, said in another way, does a dipole rotate with respect to the center ? or the start ? or the end point ? can this be implemented in a general way ? Which point do you currently use in your implementation ?
Also, wrt to this
```
the rotation transformation only changed the R1 and R2 attributes and left T1 and T2 untouched. This is true for a tilt/roll angle but not for a pitch or yaw angle
```
a dipole roll will also produce translations on the input and output reference systems T1 and T2.

SebastienJoly commented:
Thanks for the feedback @lfarv!

I will create a new file (transformation.py) to move all the new functions then. In addition, I have kept the old function as a quick way to benchmark the new implementation. Of course, the idea was to replace the old function and not to keep both in pyAT. We can also keep it in a deprecated library if needed.

Finally, thanks for bringing this to my attention. I did not realize this but you are absolutely right.

>     * I think that there is a problem related to AT's choice of coordinates (also present in the old implementation): T1[1] and T1[3] are used to model rotations around the X or Y axis. But these coordinates are px = x'(1+δ) and not x': so the angular kick x' depends on the momentum. No way to deal with that with a matrix… This should introduce an error on the dispersion generated by rotation errors. Since SC/pySC is based AT for the optics, I don't think comparing the dispersion will show anything. Is there another possible  benchmark?

Could we use Xsuite/elegant to cross check what we have and see the magnitude of the dispersion error due to the pyAT choice of coordinates?

SebastienJoly commented:
Thanks for the comments as well @oscarxblanco

> Dear @SebastienJoly , thank you very much. The documentation still misses the reference point around which the transformation is applied for curved reference systems. Or, said in another way, does a dipole rotate with respect to the center ? or the start ? or the end point ? can this be implemented in a general way ? Which point do you currently use in your implementation ?

The rotation is applied with respect to the center of the element. Based on my current understanding of the problem, while the rotation could be applied with respect to a face, we would need to change the implementation of the R1,R2 (and maybe T1,T2) elements. Is this a feature you would like to have? And if so, to study which problem?

> 
> ```
> the rotation transformation only changed the R1 and R2 attributes and left T1 and T2 untouched. This is true for a tilt/roll angle but not for a pitch or yaw angle
> ```
> 
> a dipole roll will also produce translations on the input and output reference systems T1 and T2.

I am not so sure. The way I picture this is, for a roll which is a rotation along the Z-axis, the entrance and exit faces of the element are simply rotated. By doing so, no additional displacement is added in any plane. Otherwise, it would mean that the use of skew elements also require T1 and T2 vectors.
On the contrary, a pitch or yaw will shift the entrance and exit faces; therefore, T1 and T2 are required.

lfarv commented:
@oscarxblanco, @SebastienJoly 

> The rotation is applied with respect to the center of the element

In my understanding, the rotation is around an axis joining the entry and exit points of the bending magnet. This seems to me the easiest description of bending magnet rotation.

SebastienJoly commented:
Yes, this is actually a better description of the rotation reference point in a dipole or any element.

oscarxblanco commented:
> > a dipole roll will also produce translations on the input and output reference systems T1 and T2.
> 
> I am not so sure. The way I picture this is, for a roll which is a rotation along the Z-axis, the entrance and exit faces of the element are simply rotated. By doing so, no additional displacement is added in any plane. Otherwise, it would mean that the use of skew elements also require T1 and T2 vectors. On the contrary, a pitch or yaw will shift the entrance and exit faces; therefore, T1 and T2 are required.

For a dipole, due to the curvature, there will be an angle between the magnet face and the plane perpendicular to the sagitta around which you and @lfarv propose to define the roll. You will need T1 and T2 to describe that angle.

lfarv commented:
@oscarxblanco:
Yes, rotating around the straight line joining entrance and exit requires a rotation at both faces around all 3 axis, which is handled by T1[1], T1[3] and rotation R1 (no translation), and similarly for exit. My point is that T1[1] and t1[3] are not true rotations, because of the dependence on $\delta$.

oscarxblanco commented:
> @oscarxblanco: Yes, rotating around the straight line joining entrance and exit requires a rotation at both faces around all 3 axis, which is handled by T1[1], T1[3] and rotation R1 (no translation), and similarly for exit. My point is that T1[1] and t1[3] are not true rotations, because of the dependence on δ .

Dear @lfarv , I have no easy solution or suggestion for this, and it is worth knowing. Could this be included as a note on the help message ? For example,
"AT describes the ultra-relativistic beam dynamics in 6D phase space coordinates, which differ from 3D spatial angles in an expansion w.r.t. energy to first order by a factor $(1+\delta)$,  where $\delta$ is the relative energy offset. In general this introduces a small angle error, but could create an undesired effect for large energy offsets."

Also, I think the function `rotate_traslate_element` should be renamed.

SebastienJoly commented:
I think I am beginning to understand now. Yet, I have no idea how to treat this problem with this implementation, apart from mentioning it in the documentation.
Do you have any comment/suggestion with the new function?
Otherwise, this is my plan:
1. Move the new function to transformation.py and rename it to rotate_elem()
2. Move the previous function to deprecated.py
3. Update the docstrings w.r.t the problem with T1[1] and T1[3] elements

Am I missing something?

oscarxblanco commented:
@SebastienJoly , there is one more thing on which I have to insist.
The dipole roll requires T1 and T3.

oscarxblanco commented:
I'm testing your function, @SebastienJoly, I have made the following python script
```python
import at
import numpy
qf = at.Quadrupole('qf',0.1,3)
at.rotate_translate_elem(qf,pitch=500e-6)
print(qf.T1)
print(qf.T2)
```
And I get the following
```python
>>> exec(open('test_sjoly_transformations.py').read())
[0.     0.     0.     0.0005 0.     0.    ]
[ 0.00000000e+00  0.00000000e+00 -5.00000042e-05 -4.99999979e-04
  0.00000000e+00  1.25000013e-08]
```
Which does not match your documentation. This is correct if the transformation is applied at the the element entrance, in which case I expect
```matlab
T1theo = [0 0 0 tan(apitch) 0 0]
T2theor = [0 0 -magLength*tan(apitch) -sin(apitch) 0 magLength*(sec(apitch)-1)]
```
The case you describe in your documentation is a rotation around the mid point, and the theoretical T1 and T2 are:
```matlab
T1theomidpoint=[   0 0 -magLength/2*(sin(apitch)) tan(apitch) 0 magLength/2*(1-cos(apitch))]
T2theomidpoint=[    0 0 -magLength/2*tan(apitch) -sin(apitch) 0 magLength/2*(sec(apitch)-1)]
```
You could do a numerical check in SC matlab using the following lines
```matlab
format long
dx = 0;
dy = 0;
dz = 0;
atilt = 0.0;
apitch = 0.5e-3;
ayaw = 0.0;
ax = apitch;
ay = ayaw;
az = atilt;
magTheta = 0;
magLength = 0.1;
[T1,T2,R1,R2] = SCgetTransformation(dx,dy,dz,ax,ay,az,magTheta,magLength,'refPoint','entrance');
%[T1,T2,R1,R2] = SCgetTransformation(dx,dy,dz,ax,ay,az,magTheta,magLength,'refPoint','center');
T1andT2fromSC=[T1';T2']
R1
R2
```
Could you please clarify ?

lfarv commented:
@SebastienJoly and @oscarxblanco 

I made another test on a tile of a bending magnet:
```
>>> dip=at.Dipole("ABCD", 1.0, 0.8)
>>> at.rotate_translate_elem(dip, tilt=0.01)
>>> print(dip)
Dipole:
       FamName: ABCD
        Length: 1.0
    PassMethod: BndMPoleSymplectic4Pass
  BendingAngle: 0.8
 EntranceAngle: 0.0
     ExitAngle: 0.0
      MaxOrder: 0
   NumIntSteps: 10
      PolynomA: [0. 0.]
      PolynomB: [0. 0.]
            R1: [[ 0.99995     0.          0.00999983  0.          0.          0.        ]
 [ 0.          0.99995     0.          0.00999983  0.          0.        ]
 [-0.00999983 -0.          0.99995     0.          0.          0.        ]
 [ 0.         -0.00999983  0.          0.99995     0.          0.        ]
 [ 0.          0.          0.          0.          1.          0.        ]
 [-0.         -0.         -0.         -0.          0.          1.        ]]
            R2: [[ 9.99975730e-01  1.35979871e-05 -6.96713024e-03 -9.47412466e-08
   0.00000000e+00  0.00000000e+00]
 [ 0.00000000e+00  9.99975730e-01  0.00000000e+00 -6.96695098e-03
   2.49891318e-05  0.00000000e+00]
 [ 6.96713024e-03  9.47412466e-08  1.00000146e+00  1.35983370e-05
   0.00000000e+00  0.00000000e+00]
 [ 0.00000000e+00  6.96695098e-03  0.00000000e+00  9.99950000e-01
  -7.17344135e-03  0.00000000e+00]
 [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00
   1.00000000e+00  0.00000000e+00]
 [-2.49897748e-05 -3.39818883e-10 -7.17362593e-03 -9.75492405e-08
   0.00000000e+00  1.00000000e+00]]
            T1: [0. 0. 0. 0. 0. 0.]
            T2: [ 1.32068842e-05  2.49891318e-05 -3.79120049e-03 -7.17344135e-03
  0.00000000e+00  1.35983172e-05]
             K: 0.0
>>> 
```
Since T1 = 0, we are left with only an S rotation at entrance: this is clearly a **tilt applied to the entrance face**, not on the magnet centre. Consequently, the exit is out of the mid plane, and we get a y translation at the exit (`T2[2]= -3.79e-3`). This is possibly correct, but not what we expected. I think that the rotation around the centre (or middle of the sagitta for a bending magnet) would be much more convenient. Then for the tilt of a bending magnet, I expect something more symmetrical:

`T1[0] = T1[2] = T2[0] = T2[2] = 0`: no displacement
`T1[1] = ± T2[1]`: equal or opposite X rotations (I can't guess the sign)
`T1[3] = ± T2[3]`
`R1` and `R2` opposite rotations

SebastienJoly commented:
Thanks for pointing out the discrepancy between the midpoint in the docstrings and the one implemented. It turns out I did not fully understand the implementation. You were right, the midpoint was set on the entrance face.
I heavily modified the function to be closer to the one available in SC. We now have the possibility to choose where to set the midpoint (center or entrance) and also to apply transverse shifts (dx/dy).
We still have the possibility to apply relative rotations with both midpoint options. I just needed to store the rotation matrix in the element attributes, is it ok?
However, the transverse shifts cannot be preserved with the relative argument. It becomes too complex when rotations and shifts are present at the same time as the transformations are not commutative.
I spent some time reworking the docstrings and adding many comments pointing out to the equations I used from the paper from Marco Venturini and Thorsten Hellert.
I moved all the rotation functions in transformation.py and called the main function transform_elem.
I reshuffled the utils.py by moving the old rotation function to deprecated.py and changed the set_tilt, set_rotation, tilt_elem to call transform_elem.

The new function passed all the tests you listed. Unfortunately, I do not have access to matlab on my laptop so I cannot make more tests using SC. I will try to install matlab next week.

Is there something else that needs to be fixed or is it ready for review?

lfarv commented:
```
>>> at.transform_elem(dip, tilt=0.01)
>>> print(dip)
Dipole:
       FamName: ABCD
        Length: 1.0
    PassMethod: BndMPoleSymplectic4Pass
  BendingAngle: 0.8
 EntranceAngle: 0.0
     ExitAngle: 0.0
      MaxOrder: 0
   NumIntSteps: 10
      PolynomA: [0. 0.]
      PolynomB: [0. 0.]
            R1: [[ 9.99957582e-01  3.63825078e-22  9.21052627e-03  3.35116259e-24
   0.00000000e+00  0.00000000e+00]
 [ 0.00000000e+00  9.99957583e-01  0.00000000e+00  9.21045643e-03
  -1.79337528e-05  0.00000000e+00]
 [-9.21052627e-03 -3.35116259e-24  9.99965165e-01  3.63827837e-22
   0.00000000e+00  0.00000000e+00]
 [ 0.00000000e+00 -9.21045643e-03  0.00000000e+00  9.99950000e-01
  -3.89411852e-03  0.00000000e+00]
 [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00
   1.00000000e+00  0.00000000e+00]
 [ 1.79338888e-05  6.52507528e-27 -3.89414805e-03 -1.41684882e-24
   0.00000000e+00  1.00000000e+00]]
            R2: [[ 9.99957582e-01  0.00000000e+00 -9.21052627e-03 -0.00000000e+00
   0.00000000e+00  0.00000000e+00]
 [ 0.00000000e+00  9.99957583e-01  0.00000000e+00 -9.21045643e-03
   1.79337528e-05  0.00000000e+00]
 [ 9.21052627e-03  0.00000000e+00  9.99965165e-01  0.00000000e+00
   0.00000000e+00  0.00000000e+00]
 [ 0.00000000e+00  9.21045643e-03  0.00000000e+00  9.99950000e-01
  -3.89411852e-03  0.00000000e+00]
 [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00
   1.00000000e+00  0.00000000e+00]
 [-1.79338888e-05 -0.00000000e+00 -3.89414805e-03 -0.00000000e+00
   0.00000000e+00  1.00000000e+00]]
            T1: [-2.77555756e-17  1.79338888e-05 -2.21255173e-19 -3.89414805e-03
  0.00000000e+00 -1.41059322e-37]
            T2: [ 4.80618242e-21  1.79337528e-05 -5.21791960e-19 -3.89411852e-03
  0.00000000e+00  0.00000000e+00]
          _r3d: [[ 9.99957583e-01 -9.21045643e-03 -1.79337528e-05]
 [ 9.21045643e-03  9.99950000e-01  3.89411852e-03]
 [-1.79337528e-05 -3.89411852e-03  9.99992418e-01]]
             K: 0.0
```
This is now conform to my expectations:

`T1[0] ≈ T1[2] ≈ T2[0] ≈ T2[2] ≈ 0`: no displacement
`T1[1] = T2[1]`: equal X rotations (I can't guess the sign)
`T1[3] = T2[3]`; equal Y rotations
`R1` and `R2` opposite rotations

lfarv commented:
> I just needed to store the rotation matrix in the element attributes, is it ok?

If I understand well, it's just used to revert to the initial angles, for the relative option? No problem for me.

lfarv commented on pyat/at/lattice/transformation.py:
```diff
@@ -0,0 +1,280 @@
+import numpy
+#import re
+from typing import Union, Tuple, List, Type
```
Adding this line on top will ensure compatibility with old python version (3.7 for us):
```python
from __future__ import annotations
```
Also, `Union, Tuple, List, Type` are not used, you can remove this line.

lfarv commented on pyat/at/lattice/transformation.py:
```diff
@@ -0,0 +1,280 @@
+    if relative:
+        if hasattr(elem, '_r3d'):
+            if midpoint == 'center':
+                r3d = numpy.dot(RB_half.T, numpy.dot(elem._r3d, RB_half))
```
What about:
```python
r3d = RB_half.T @ elem._r3d @ RB_half
```
It looks more readable for me…

lfarv commented on pyat/at/lattice/transformation.py:
```diff
@@ -0,0 +1,280 @@
+    
+    # R2, T2
+    # XYZ - axes unit - vectors expressed in the xyz coordinate system
+    X_axis = numpy.dot(RB, x_axis)
+    Y_axis = numpy.dot(RB, y_axis)
```
Same thing here

lfarv commented on pyat/at/lattice/transformation.py:
```diff
@@ -0,0 +1,280 @@
+    pyAT describes the ultra-relativistic beam dynamics in 6D phase space 
+    coordinates, which differ from 3D spatial angles in an expansion with 
+    respect to the energy to first order by a factor (1 + $\\delta$) , where 
+    $\\delta$ is the relative energy offset. In general this introduces a small 
+    angle error, but could create an undesired effect for large energy offsets.
```
It introduces a (hopefully small)  **spurious dispersion**: it's an angle proportional to the off-momentum. I would rephrase it that way…

lfarv commented:
@SebastienJoly : I get many PEP8 errors. I suggest that you run [flake8](https://flake8.pycqa.org/en/latest/#) to list the errors and [black](https://black.readthedocs.io/en/stable/) to reformat the code,

lfarv commented on pyat/at/lattice/transformation.py:
```diff
@@ -0,0 +1,280 @@
+    *r3d* matrix and incremented by the input arguments.
+    *relative* only allows to add the previous angles, not the transverse 
+    shifts.
+    The shift is always conserved regardless of the value of *relative*.
```
> The shift is always conserved regardless of the value of *relative*.

This sentence is not clear. Could it be "The shift is always absolute regardless of the value of *relative*"?

lfarv commented:
@SebastienJoly, @oscarxblanco, @swhite2401 
```
    Parameters:
        ...
        midpoint:       Midpoint reference (entrance/center)
        ...
```
I have a few remarks on this argument, and I would like to have the opinion of other people on that:
- the choice of `midpoint` for the argument name: it could be `reference`, or `origin`,
- rather the strings, we are usually using enumerations for that: this avoids misspellings. See [cavity_access.py](https://at.readthedocs.io/en/latest/p/api/at.physics.energy_loss.html#at.physics.energy_loss.ELossMethod)
- I prefer "centre" to "center" (U.K. English). See in MAD-X sequences "..., REFER=CENTRE,...". However, it seems that there are already some diverging options in AT…

lfarv commented:
@SebastienJoly: It seems that using R1 and R2 you can also apply longitudinal positioning errors. Isn't it worth adding a `ds` or `dz` argument to allow that?

lfarv commented on pyat/at/lattice/transformation.py:
```diff
@@ -0,0 +1,280 @@
+    
+    # Extract current transformations if relative=True
+    tilt0, pitch0, yaw0 = 0.0, 0.0, 0.0
+    t10, t20 = numpy.zeros(6), numpy.zeros(6)
```
`t10`, `t20` are never used. Can this be removed?

SebastienJoly replied:
Yes, I forgot to remove them. It was when I tried to allow both relative rotations and translations.

SebastienJoly commented:
I have modified the functions to account for all your comments @lfarv 
Thanks a lot for taking the time to go through all of it!
I am not so familiar with the Enum library in python, feel free to correct what I have done. And @oscarxblanco @swhite2401, on what convention do we agree for the reference on which we apply the transformation? centre/entrance?

lfarv commented on pyat/at/lattice/transformation.py:
```diff
@@ -0,0 +1,315 @@
+    )
+
+
+class reference_point(Enum):
```
Following PEP8, class names should follow the CapWords convention. So here it could be:
```python
class Reference(Enum):
# or
class AtRefer(Enum):
```

oscarxblanco commented:
> Hello @SebastienJoly
> 
> > it is not possible in pyAT to longitudinally shift an element. I do not know what would be the different to apply such a shift with R1, R2 matrices instead of T1, T2.
> 
> You cannot shift an element longitudinally with T1, T2. But R1 and R2 can represent any transfer matrix. For a straight element, if you put in R1 a small drift matrix and in R2 the same negative drift, it represents a longitudinal translation of the element.
> 
> I made a test with your last version:
> 
> ```
> >>> import at
> >>> qd=at.Quadrupole("QP", 1.0, 0.5)
> >>> at.transform_elem(qd, dz=0.02)
> >>> print(qd)
> Quadrupole:
>        FamName: QP
>         Length: 1.0
>     PassMethod: StrMPoleSymplectic4Pass
>       MaxOrder: 1
>    NumIntSteps: 10
>       PolynomA: [0. 0.]
>       PolynomB: [0.  0.5]
>             R1: [[ 1.    0.02 -0.   -0.    0.    0.  ]
>  [ 0.    1.    0.    0.    0.    0.  ]
>  [-0.   -0.    1.    0.02  0.    0.  ]
>  [ 0.    0.    0.    1.    0.    0.  ]
>  [ 0.    0.    0.    0.    1.    0.  ]
>  [-0.   -0.   -0.   -0.    0.    1.  ]]
>             R2: [[ 1.   -0.02 -0.    0.    0.    0.  ]
>  [ 0.    1.    0.    0.    0.    0.  ]
>  [-0.    0.    1.   -0.02  0.    0.  ]
>  [ 0.    0.    0.    1.    0.    0.  ]
>  [ 0.    0.    0.    0.    1.    0.  ]
>  [-0.    0.   -0.    0.    0.    1.  ]]
>             T1: [0.   0.   0.   0.   0.   0.02]
>             T2: [ 0.    0.    0.    0.    0.   -0.02]
>           _r3d: [[1. 0. 0.]
>  [0. 1. 0.]
>  [0. 0. 1.]]
>              K: 0.5
> ```
> 
> This looks almost correct: R1 and R2 are indeed drift matrices. But I do not understand T1[5] and T2[5]: the 6th coordinate is the path **lengthening with respect to the reference particle**. A additional drift should not contribute to that. It's as if this coordinate was the absolute path length and not the lengthening with respect to the reference. Here it does not matter too much because T1 and T2 cancel out, but have a doubt…

@lfarv , I consider T1[5] and T2[5] correct as they are implemented now by @SebastienJoly. In a straight magnet they will simply cancel out. For an RF cavity it will create a small phase offset when traversing the cavity. For a bending magnet the longitudinal displacement produces a small transverse offset and a shortened longitudinal displacement.

The sign also is correct as dz points in  the opposite direction of the longitudinal coordinate $c\tau$.

None of these will add to the total path lengthening of the particle trajectory.

oscarxblanco commented:
> I have modified the functions to account for all your comments @lfarv Thanks a lot for taking the time to go through all of it! I am not so familiar with the Enum library in python, feel free to correct what I have done. And @oscarxblanco @swhite2401, on what convention do we agree for the reference on which we apply the transformation? centre/entrance?

@SebastienJoly , centre and entrance seem OK to me. But you are free to choose.

oscarxblanco commented:
@SebastienJoly ,
I want comment about the implementation of the 'relative' argument.

I think the code could be more sustainable and understandable if is divided in pieces.
1) Transform from dx,dy,dz,ax,ay,az to T1,T2,R1,R2
2) Do the inverse, i.e. transform T1,T2,R1,R2 to dx,dy,dz,ax,ay,az (or in the current case just the angles from your implementation)
3) Now you could use either directly function 1) when doing absolute transformations; alternatively go through function 2) and add the displacements to be input to function 1) for relative displacements.

You gain a bit of flexibility by being able to recover the current state, and modified, instead of adding some values calculated inside the function that you would need to carefully inspect in case of problems.

SebastienJoly commented:
@oscarxblanco 
Thank you for the idea. I implemented your suggestion, and now I can use the relative argument without the _r3d attribute, both with shifts and rotations! I tested with a reference point set at the entrance or centre of the element and it is working in both scenarios.
By comparing the R1, R2, T1 and T2 of a full rotation/shift or two half rotations/shifts, I retrieve exactly the same results.
Now, the results are almost equivalent when I am comparing a full rotation+shift or two half rotation+shift. I get an error of the order of 1e-6. I suspect a loss of information when computing the matrix inverse.

As the relative argument works with shifts and rotations, I modified the shift_elem() function in utils.py and its tests as the new function allows longitudinal shifts and the effect of a bending angle on R1, R2, T1 and T2. Is it fine with @lfarv @swhite2401 ?

lfarv commented on pyat/at/lattice/utils.py:
```diff
@@ -1094,25 +1012,28 @@ def set_tilt(ring: Sequence[Element], tilts, relative=False) -> None:
 
 
-def set_shift(ring: Sequence[Element], dxs, dzs, relative=False) -> None:
+def set_shift(ring: Sequence[Element], dxs, dys, dzs, relative=False) -> None:
```
This should be
```python
def set_shift(ring: Sequence[Element], dxs, dys, dzs=0.0, *, relative=False) -> None:
```
You have to make `dzs` a keyword argument with default 0.0, otherwise you break the compatibility for cases where it is called with only three positional arguments (`set_shift(ring, dxs, dys)`).

In addition, making `relative` keyword only avoids confusion when using 4 positional arguments.

lfarv commented on pyat/test/test_lattice_utils.py:
```diff
@@ -199,7 +199,7 @@ def test_tilt_elem():
     elem = elt.Drift("Drift", 1.0)
     # Test shift_elem function
-    shift_elem(elem, 1.0, 0.5)
+    shift_elem(elem, 1.0, 0.5, 0.0)
```
`shift_elem` has no entry for  longitudinal displacement. So the 4th argument is interpreted here as `relative=0.0`.

To ensure the backward compatibility, you have to make sure that `test_lattice_utils` still runs unchanged. The only thing you can do is to add new tests for the new features. But the old ones should not be modified.

SebastienJoly replied:
The backward compatibility is now preserved with the new functions

lfarv commented on pyat/at/lattice/transformation.py:
```diff
@@ -0,0 +1,390 @@
+            yaw0 = np.arcsin(r3d[0, 2])
+            pitch0 = np.arctan2(-r3d[1, 2], r3d[2, 2])
+
+        else:
+            print(
+                "Cannot apply a relative transformation without a prior"
+                "transformation, assume an absolute transformation instead."
+            )
```
Can't we assume that if T1, R1 are absent, it just means that they are 0 ?

For instance:
```python
try:
    oldR1 = elem.R1
except AttributeError:
    oldR1 = np.identity(6)
try:
    oldT1 = elem.T1
except AttributeError:
    oldT1 = np.zero(6)
```
This fixes the case where you have one and not the other: in you implementation, you assume that both are zero .

SebastienJoly replied:
I do not think my implementation allows to have either T1 or R1. It sets both quantities the first time the function is called, even if the transformation results in a zero vector for T1 and an identity matrix for R1. Same thing for T2 and R2.

lfarv replied:
Now, I don't think the `print` statement is useful: no attributes means the initial state is zero:
You could just remove the` else:` clause.

If you find the information really useful, then it should be a `warning` rather than a `print`.

SebastienJoly replied:
The print was more of a sanity check but I agree it is not crucial
I removed it

SebastienJoly reviewed (commented):
I applied the discussed changes, you can find them in my different answers to your comments

lfarv commented:
Could you put the stars in the argument lists:
```python
def shift_elem(elem: Element, dx: float = 0.0, dy: float = 0.0, dz: float = 0.0, *, relative: bool = False) -> None:
...
def set_shift(ring: Sequence[Element], dxs, dys, dzs=0.0, *, relative=False) -> None:
```
This makes `relative` a keyword-only argument, which avoids confusion with a possible `dz` positional argument.

lfarv commented:
Hello @SebastienJoly:

I found another problem: Using `set_rotation` in absolute mode (`relative=False`, the default) resets all the translations to zero. And similarly, `set_shift` resets all rotations to zero. 

Practically, it means that only the "relative" mode should be used to be able to set both translations and rotations. Or one should use directly the `transform_elem` function.

Ideally, all elementary functions like `shift_elem`, `tilt_elem` …, and the `Element` properties like `Element.dx`,… should go through `transform_elem`, which should have new default values: `dx=None,…` instead of `dx=0.0,…`, meaning "keep the value unchanged". Whatever `relative` is, one needs to reverse-engineer the current angles and translations, and then apply the explicitly specified values as relative or absolute modifications. Since this is a new function, we can still modify its default arguments.

This is still a big modification, it turns out that this PR goes much further than the implementation of the right formulas… What is your advice, @SebastienJoly, @swhite2401 and @oscarxblanco ?

Otherwise, we could temporarily take this PR as it is, with a word of caution, and leave the reorganisation for a second step?

SebastienJoly commented on pyat/at/lattice/elements.py:
```diff
@@ -483,6 +483,15 @@ def dy(self) -> float:
+
+    @dz.setter
+    def dz(self, value: float) -> None:
+        self._setshift(-value, 5)
```
You are absolutely right, my mistake.

SebastienJoly commented:
> Hello @SebastienJoly:
> 
> I found another problem: Using `set_rotation` in absolute mode (`relative=False`, the default) resets all the translations to zero. And similarly, `set_shift` resets all rotations to zero.
> 
> Practically, it means that only the "relative" mode should be used to be able to set both translations and rotations. Or one should use directly the `transform_elem` function.
> 
> Ideally, all elementary functions like `shift_elem`, `tilt_elem` …, and the `Element` properties like `Element.dx`,… should go through `transform_elem`, which should have new default values: `dx=None,…` instead of `dx=0.0,…`, meaning "keep the value unchanged". Whatever `relative` is, one needs to reverse-engineer the current angles and translations, and then apply the explicitly specified values as relative or absolute modifications. Since this is a new function, we can still modify its default arguments.
> 
> This is still a big modification, it turns out that this PR goes much further than the implementation of the right formulas… What is your advice, @SebastienJoly, @swhite2401 and @oscarxblanco ?
> 
> Otherwise, we could temporarily take this PR as it is, with a word of caution, and leave the reorganisation for a second step?

I did not consider this, but it makes sense. Now that the rotations and shifts are handled together, it complicates the overall picture. 
I agree that in the best case scenario, the properties `Element.dx`, `Element.dy`, and `Element.dz` should be handled by `transform_elem`. In addition, there is now the possibility to reverse engineer the offsets and angles with the function. However, it feels like a considerable amount of work... Unfortunately, I will not be able to dedicate much more time on the function until the beginning of summer.
I vote for taking the PR as it is and investing more time in its reorganization later.

lfarv commented:
@SebastienJoly and @oscarxblanco : I agree with both of you, let's merge this PR and acknowledge the very nice work done be @SebastienJoly. All the rest is a rather simple reorganisation which I can handle in another PR (no need touch all the computations).

lfarv commented:
@SebastienJoly : the master branch has been updated and there is now a conflict in `utils.py`. I'll solve this and let @oscarxblanco or you approve the final state.

lfarv commented:
@SebastienJoly, @oscarxblanco : the conflicts in `utils.py` are now solved. You can check in the "modified files" above that your changes to `utils.py` are intact. Feel free to approve (I can't do it) and merge.

T-Nicholls commented:
[like]  Nicholls, Tobyn (DLSLtd,RAL,LSCI) reacted to your message:
________________________________
From: Sébastien Joly ***@***.***>
Sent: Monday, April 14, 2025 6:23:07 AM
To: atcollab/at ***@***.***>
Cc: Subscribed ***@***.***>
Subject: Re: [atcollab/at] Implementation of a new function to rotate an element (PR #921)


Thanks for the comments as well @oscarxblanco<https://github.com/oscarxblanco>

Dear @SebastienJoly<https://github.com/SebastienJoly> , thank you very much. The documentation still misses the reference point around which the transformation is applied for curved reference systems. Or, said in another way, does a dipole rotate with respect to the center ? or the start ? or the end point ? can this be implemented in a general way ? Which point do you currently use in your implementation ?

The rotation is applied with respect to the center of the element. Based on my current understanding of the problem, while the rotation could be applied with respect to a face, we would need to change the implementation of the R1,R2 (and maybe T1,T2) elements. Is this a feature you would like to have? And if so, to study which problem?

the rotation transformation only changed the R1 and R2 attributes and left T1 and T2 untouched. This is true for a tilt/roll angle but not for a pitch or yaw angle


a dipole roll will also produce translations on the input and output reference systems T1 and T2.

I am not so sure. The way I picture this is, for a roll which is a rotation along the Z-axis, the entrance and exit faces of the element are simply rotated. By doing so, no additional displacement is added in any plane. Otherwise, it would mean that the use of skew elements also require T1 and T2 vectors.
On the contrary, a pitch or yaw will shift the entrance and exit faces; therefore, T1 and T2 are required.

—
Reply to this email directly, view it on GitHub<https://github.com/atcollab/at/pull/921#issuecomment-2800591795>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AJVXGO2AAZTB3UYQ4JIWSST2ZNH4XAVCNFSM6AAAAAB23HVONWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMBQGU4TCNZZGU>.
You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

[https://avatars.githubusercontent.com/u/56296618?s=20&v=4]SebastienJoly left a comment (atcollab/at#921)<https://github.com/atcollab/at/pull/921#issuecomment-2800591795>

Thanks for the comments as well @oscarxblanco<https://github.com/oscarxblanco>

Dear @SebastienJoly<https://github.com/SebastienJoly> , thank you very much. The documentation still misses the reference point around which the transformation is applied for curved reference systems. Or, said in another way, does a dipole rotate with respect to the center ? or the start ? or the end point ? can this be implemented in a general way ? Which point do you currently use in your implementation ?

The rotation is applied with respect to the center of the element. Based on my current understanding of the problem, while the rotation could be applied with respect to a face, we would need to change the implementation of the R1,R2 (and maybe T1,T2) elements. Is this a feature you would like to have? And if so, to study which problem?

the rotation transformation only changed the R1 and R2 attributes and left T1 and T2 untouched. This is true for a tilt/roll angle but not for a pitch or yaw angle


a dipole roll will also produce translations on the input and output reference systems T1 and T2.

I am not so sure. The way I picture this is, for a roll which is a rotation along the Z-axis, the entrance and exit faces of the element are simply rotated. By doing so, no additional displacement is added in any plane. Otherwise, it would mean that the use of skew elements also require T1 and T2 vectors.
On the contrary, a pitch or yaw will shift the entrance and exit faces; therefore, T1 and T2 are required.

—
Reply to this email directly, view it on GitHub<https://github.com/atcollab/at/pull/921#issuecomment-2800591795>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AJVXGO2AAZTB3UYQ4JIWSST2ZNH4XAVCNFSM6AAAAAB23HVONWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMBQGU4TCNZZGU>.
You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

This e-mail and any attachments may contain confidential, copyright and or privileged material, and are for the use of the intended addressee only. If you are not the intended addressee or an authorised recipient of the addressee please notify us of receipt by returning the e-mail and do not use, copy, retain, distribute or disclose the information in or attached to the e-mail. Any opinions expressed within this e-mail are those of the individual and not necessarily of Diamond Light Source Ltd.
Diamond Light Source Ltd. cannot guarantee that this e-mail or any attachments are free from viruses and we cannot accept liability for any damage which you may sustain as a result of software viruses which may be transmitted in or with the message.
Diamond Light Source Limited (company no. 4375679). Registered in England and Wales with its registered office at Diamond House, Harwell Science and Innovation Campus, Didcot, Oxfordshire, OX11 0DE, United Kingdom.
